File: src/stores/memory.ts
/**
 * yaml-flow - Memory Store
 * 
 * In-memory store implementation. Works in both browser and Node.js.
 * Data is lost when the process exits - use for testing or short-lived flows.
 */

import type { FlowStore, RunState } from '../core/types.js';

export class MemoryStore implements FlowStore {
  private runs: Map<string, RunState> = new Map();
  private data: Map<string, Record<string, unknown>> = new Map();

  async saveRunState(runId: string, state: RunState): Promise<void> {
    this.runs.set(runId, { ...state });
  }

  async loadRunState(runId: string): Promise<RunState | null> {
    const state = this.runs.get(runId);
    return state ? { ...state } : null;
  }

  async deleteRunState(runId: string): Promise<void> {
    this.runs.delete(runId);
    this.data.delete(runId);
  }

  async setData(runId: string, key: string, value: unknown): Promise<void> {
    if (!this.data.has(runId)) {
      this.data.set(runId, {});
    }
    const runData = this.data.get(runId)!;
    runData[key] = value;
  }

  async getData(runId: string, key: string): Promise<unknown> {
    return this.data.get(runId)?.[key];
  }

  async getAllData(runId: string): Promise<Record<string, unknown>> {
    return { ...(this.data.get(runId) ?? {}) };
  }

  async clearData(runId: string): Promise<void> {
    this.data.delete(runId);
  }

  async listRuns(): Promise<string[]> {
    return Array.from(this.runs.keys());
  }

  /**
   * Clear all data (useful for testing)
   */
  clear(): void {
    this.runs.clear();
    this.data.clear();
  }
}

File: pycli/pylib/stores/memory.py
"""
yaml-flow - Memory Store

In-memory store implementation. Works in any Python environment.
Data is lost when the process exits - use for testing or short-lived flows.

Port of src/stores/memory.ts
"""
from __future__ import annotations


class MemoryStore:
    """In-memory implementation of the StepMachineStore protocol."""

    def __init__(self) -> None:
        self._runs: dict[str, dict] = {}      # runId -> state dict
        self._data: dict[str, dict] = {}       # runId -> {key: value}

    def save_run_state(self, run_id: str, state: dict) -> None:
        self._runs[run_id] = {**state}

    def load_run_state(self, run_id: str) -> dict | None:
        state = self._runs.get(run_id)
        return {**state} if state is not None else None

    def delete_run_state(self, run_id: str) -> None:
        self._runs.pop(run_id, None)
        self._data.pop(run_id, None)

    def set_data(self, run_id: str, key: str, value: object) -> None:
        if run_id not in self._data:
            self._data[run_id] = {}
        self._data[run_id][key] = value

    def get_data(self, run_id: str, key: str) -> object:
        run_data = self._data.get(run_id)
        if run_data is None:
            return None
        return run_data.get(key)

    def get_all_data(self, run_id: str) -> dict:
        return {**(self._data.get(run_id) or {})}

    def clear_data(self, run_id: str) -> None:
        self._data.pop(run_id, None)

    def list_runs(self) -> list[str]:
        return list(self._runs.keys())

    def clear(self) -> None:
        """Clear all data (useful for testing)."""
        self._runs.clear()
        self._data.clear()
